CAMEL-18661: Make span current and clear scope properly for async processing

lmolkova opened the pull request:
Fixesr https://issues.apache.org/jira/browse/CAMEL-18661

~~**!Looking for input on the overall approach!**~~

This change enables  current span propagation to underlying libraries and end-user-code (for OpenTelemetry):
 for all spans created by Camel, it calls into `span.makeCurrent()`

However, OpenTelemetry (and other tracing tools) rely on `ThreadLocal` to propagate context. They do it carefully for Executors, Reactor, etc, so it works in async scenarios too. 

**Instrumentations have to close scope returned by `span.makeCurrent()` on the same thread where it was created to avoid leaking context (by leaving current span on current thread).**

This is easy to guarantee with sync operations in Camel - ExchangeStarted/Sending and ExchangeCompleted/Sent events (where spans start and end) are called on the same thread.

However, for async operations, they are called on different threads. 

This change also adds a new event - `ExchangeAsyncStarted` (naming needs more work), that notifies that `Processor.process` call has ended, i.e. async operation has started. Tracers would end the scope (but not span) when this event is sent.

**Alternative approach:**

add a dependency on `camel-tracing` to `camel-base-engine` and call into `ActiveSpanManager` directly:

```java
try(AutoCloseable scope = ActiveSpanManager.makeSpanCurrent(exchange)) {
    sync = processor.process(exchange, async);
}
```

This would be easier to maintain and would more or less guarantee that scope is disposed on the same thread, however, dependency on tracing might not be desirable.


**Traces with this change**

(tick/testme spans come from Camel and were previously not correlated with ServiceBus spans coming from Azure SDK)

<img width="949" alt="image" src="https://user-images.githubusercontent.com/2347409/201442066-5d9684c4-41b5-4556-b011-3d1c443ff51d.png">

lmolkova commented:
General question: is there a good place/way to wrap callback executions (user or 3rd party library) with something like 
```
try(AutoCloseable scope = ActiveSpanManager.makeSpanCurrent(exchange)) {
    sync = processor.process(exchange, async);
}
```

davsclaus commented:
Sorry for not giving this PR attention - however we have been buys (as usual).

Observability with open-telemtry and micrometer is important and thanks for helping with this.

oscerd commented:
I'm sorry, too. It was on my to-do list to review this and I forgot about it

lmolkova commented:
@davsclaus @oscerd thank you! Once you folks have time to take a look, please let me know what approach you'd like me to pursue nd I'm happy to work on it further.

lmolkova commented:
@davsclaus @oscerd friendly ping to take a look. I'd be happy to continue working on it based on your initial feedback.

davsclaus commented on core/camel-api/src/main/java/org/apache/camel/AsyncCallback.java:
```diff
@@ -19,7 +19,7 @@
  * The callback interface for an {@link AsyncProcessor} so that it can notify you when an {@link Exchange} is done.
  * <p/>
- * For example a {@link AsyncProcessor} should invoke the done method when the {@link Exchange} is ready to be continued
+ * For example a {@link AsyzncProcessor} should invoke the done method when the {@link Exchange} is ready to be continued
```
yes its a typo

davsclaus commented:
> add a dependency on camel-tracing to camel-base-engine and call into ActiveSpanManager directly:

This is not possible, camel-base should be 100% independent on camel itself only.

davsclaus commented:
Its actually the opposite

> var res = producer.process(exchange, ac);
if the returned value == true then it was a sync call and its done

if the returned value == false, then its async call, and currently processing by another thread (done when the callback is triggered)

lmolkova commented:
@davsclaus @oscerd  thank you for the feedback. I've implemented the current approach with the extra event and added tests. Can you please take another look?

bvahdat commented:
Hi @lmolkova 

Thanks for your contribution and support to look into this issue. Looking into `build.log`, this is the reason why it has failed:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:testCompile (default-testCompile) on project camel-tracing: Compilation failure: Compilation failure: 
[ERROR] /home/runner/work/camel/camel/components/camel-tracing/src/test/java/org/apache/camel/tracing/ActiveSpanManagerTest.java:[45,23] error: cannot find symbol
[ERROR]   symbol:   method isCurrent()
[ERROR]   location: variable span of type SpanAdapter
[ERROR] /home/runner/work/camel/camel/components/camel-tracing/src/test/java/org/apache/camel/tracing/ActiveSpanManagerTest.java:[49,24] error: cannot find symbol
[ERROR]   symbol:   method isCurrent()
[ERROR]   location: variable span of type SpanAdapter
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:testCompile (default-testCompile) on project camel-tracing: Compilation failure
	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:375)
	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:351)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:171)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:163)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.mvndaemon.mvnd.builder.SmartBuilderImpl.buildProject(SmartBuilderImpl.java:178)
	at org.mvndaemon.mvnd.builder.SmartBuilderImpl$ProjectBuildTask.run(SmartBuilderImpl.java:198)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1310)
	at org.apache.maven.plugin.compiler.TestCompilerMojo.execute(TestCompilerMojo.java:183)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:370)
	... 12 common frames omitted
```

Please note that you don't have access to the pipeline's generated `build.log` but I think you should be able to reproduce it locally.

davsclaus commented:
Using the event notifier to emit a "async processing" event is tricky as there are potentially more places this can happens in the core routing engine. However they all bubble up to the base-engine in the `CamelInternalProcessor`, especially for spans that would call external systems via "SendProcessor".

lmolkova commented:
@bvahdat  thank you! it should be fixed now.

@davsclaus I removed notifications from everywhere except `CamelInternalProcessor` and yes, everything seems to be working. 

Do you have in mind any scenarios I can run to validate to make sure I didn't miss something? I tried to cover what I can think of in [CurrentSpanTests](https://github.com/apache/camel/blob/afa146b06db478d8f359d6778fd921f419a466f2/components/camel-opentelemetry/src/test/java/org/apache/camel/opentelemetry/CurrentSpanTests.java).

davsclaus commented:
@lmolkova have you tested your latest code with the AWS servicebus to see if the diagram/span looks correct

lmolkova commented:
@davsclaus, yes, I did (with Azure ServiceBus ) and yes, it works fine and looks the same as in the description.

lmolkova commented:
I can't repro build error locally and would appreciate if someone can post log from the pipeline run. Thanks!

bvahdat commented:
> I can't repro build error locally and would appreciate if someone can post log from the pipeline run. Thanks!

Locally I get the following test failures:

```
[ERROR] Failures: 
[ERROR] org.apache.camel.management.AddEventNotifierTest.testAddAndRemove
[ERROR]   Run 1: AddEventNotifierTest.testAddAndRemove:70 expected: <8> but was: <11>
[ERROR]   Run 2: AddEventNotifierTest.testAddAndRemove:70 expected: <8> but was: <11>
[ERROR]   Run 3: AddEventNotifierTest.testAddAndRemove:70 expected: <8> but was: <11>
[INFO] 
[ERROR] org.apache.camel.management.JmxNotificationEventNotifierTest.testExchangeDone
[ERROR]   Run 1: JmxNotificationEventNotifierTest.testExchangeDone:80 Get a wrong number of events ==> expected: <8> but was: <11>
[ERROR]   Run 2: JmxNotificationEventNotifierTest.testExchangeDone:80 Get a wrong number of events ==> expected: <8> but was: <11>
[ERROR]   Run 3: JmxNotificationEventNotifierTest.testExchangeDone:80 Get a wrong number of events ==> expected: <8> but was: <11>
[INFO] 
[ERROR] org.apache.camel.management.JmxNotificationEventNotifierTest.testExchangeFailed
[ERROR]   Run 1: JmxNotificationEventNotifierTest.testExchangeFailed:107 Get a wrong number of events ==> expected: <4> but was: <6>
[ERROR]   Run 2: JmxNotificationEventNotifierTest.testExchangeFailed:107 Get a wrong number of events ==> expected: <4> but was: <6>
[ERROR]   Run 3: JmxNotificationEventNotifierTest.testExchangeFailed:107 Get a wrong number of events ==> expected: <4> but was: <6>
[INFO] 
[ERROR] org.apache.camel.management.PublishEventNotifierTest.testExchangeDone
[ERROR]   Run 1: PublishEventNotifierTest.testExchangeDone:57->ContextTestSupport.assertMockEndpointsSatisfied:376 mock://event Received message count. Expected: <9> but was: <12>
[ERROR]   Run 2: PublishEventNotifierTest.testExchangeDone:57->ContextTestSupport.assertMockEndpointsSatisfied:376 mock://event Received message count. Expected: <9> but was: <12>
[ERROR]   Run 3: PublishEventNotifierTest.testExchangeDone:57->ContextTestSupport.assertMockEndpointsSatisfied:376 mock://event Received message count. Expected: <9> but was: <12>
[INFO] 
[ERROR] org.apache.camel.management.PublishEventNotifierTest.testExchangeFailed
[ERROR]   Run 1: PublishEventNotifierTest.testExchangeFailed:72->ContextTestSupport.assertMockEndpointsSatisfied:376 mock://event Received message count. Expected: <5> but was: <7>
[ERROR]   Run 2: PublishEventNotifierTest.testExchangeFailed:72->ContextTestSupport.assertMockEndpointsSatisfied:376 mock://event Received message count. Expected: <5> but was: <7>
[ERROR]   Run 3: PublishEventNotifierTest.testExchangeFailed:72->ContextTestSupport.assertMockEndpointsSatisfied:376 mock://event Received message count. Expected: <5> but was: <7>
[INFO] 
[ERROR] org.apache.camel.management.RemoveEventNotifierTest.testRemove
[ERROR]   Run 1: RemoveEventNotifierTest.testRemove:69 expected: <17> but was: <20>
[ERROR]   Run 2: RemoveEventNotifierTest.testRemove:69 expected: <17> but was: <20>
[ERROR]   Run 3: RemoveEventNotifierTest.testRemove:69 expected: <17> but was: <20>
[INFO] 
[INFO] 
[ERROR] Tests run: 310, Failures: 6, Errors: 0, Skipped: 1 
```

Can you try running all the tests starting from the [core](https://github.com/apache/camel/tree/main/core) folder?

davsclaus commented:
also run component/camel-spring-xml as it has a bunch of tests and some about event notifier

lmolkova commented:
@bvahdat thanks! @davsclaus working on it and should will be able to fix all failures ASAP.

davsclaus commented:
[ERROR] Failures:
[ERROR] org.apache.camel.management.RemoveEventNotifierTest.testRemove
[ERROR]   Run 1: RemoveEventNotifierTest.testRemove:79 expected: <17> but was: <20>
[ERROR]   Run 2: RemoveEventNotifierTest.testRemove:79 expected: <17> but was: <20>
[ERROR]   Run 3: RemoveEventNotifierTest.testRemove:79 expected: <17> but was: <20>

davsclaus commented:
The test class needs to be renamed,

CurrentSpanTests -> CurrentSpanTest

Then when you run tests it will fail (also by CI as it was not run before)

[ERROR] Errors:
[ERROR] org.apache.camel.opentelemetry.CurrentSpanTest.testContextDoesNotLeak
[ERROR]   Run 1: CurrentSpanTest.testContextDoesNotLeak » CamelExecution Exception occurred dur...
[ERROR]   Run 2: CurrentSpanTest.testContextDoesNotLeak » CamelExecution Exception occurred dur...
[ERROR]   Run 3: CurrentSpanTest.testContextDoesNotLeak » CamelExecution Exception occurred dur...

davsclaus commented:
@lmolkova okay so there is a simple test error in core/camel-management
And then there is a test error in your new test class (need to rename the class)

                            // context is cleaned up
                            assertFalse(Span.current().getSpanContext().isValid());

This ^^^ fails

davsclaus reviewed (changes requested):
need to fix tests

davsclaus commented:
It happens when its the JDK fork-join pool that are in use

```
pool-17-thread-4 -> Valid context: false
pool-17-thread-8 -> Valid context: false
pool-17-thread-2 -> Valid context: false
pool-17-thread-9 -> Valid context: false
pool-17-thread-1 -> Valid context: false
Camel (camel-8) thread #11 - Delay -> Valid context: false
ForkJoinPool.commonPool-worker-5 -> Valid context: true
```

lmolkova commented:
thanks a lot, @davsclaus ! I was able to reproduce `testContextDoesNotLeak` locally with Java 11 (does not repro with 17 I used before)  and working on fixing it.   I'll try my best to fix it today.

davsclaus commented:
Thanks for looking into this. Yeah I only tried with JDK11, as I use that primary due to Camel v3 is still Java 11 based. 

Could be that the JDK ForJoin pool behaves differently or something?

bvahdat commented:
@lmolkova by any chance, could you please rebase on top of the main branch when you are done?

davsclaus commented:
Yeah in the mean time we have upgraded: upgrade opentelemetry to the version 1.21.0

lmolkova commented:
@bvahdat sure, I rebased!

I believe (again) that I fixed tests. There were two issues:
1. We still need to clean the context in `SendProcessor`:  since we have multiple nested spans, we need to clean scope after them.  I *think* I found all the new event-related tests.
2. Test was not correct: sometimes aggregation can be scheduled on the same thread on which outer span is still active - it's ok and if we had OTel executor instrumentation enabled, it'd populate this context on other threads too. Added check and description to the test. This behavior was different on Java 11 and Java 17 and affected Java 11 much more. It's stable now on both versions.

trask commented:
thanks @lmolkova and camel team!!

lmolkova commented:
Thank you all!
